src/smt/mod.rs

```diff
@@ -101,31 +126,69 @@ impl<'ctx> SmtCtx<'ctx> {
                         {
                             translate.push();
+                            translate.set_fuel_context(FuelContext::Body);
                             let app_z3 = translate.t_symbolic(&app);
                             if let Some(invariant) = app_z3.smt_invariant() {
                                 axioms.push((
                                     func.name, // TODO: create a new name for the axiom
                                     translate.local_scope().forall(&[], &invariant),
                                 ));
                             }
+                            translate.set_fuel_context(FuelContext::Call);
+                            translate.pop();
+                        }
+
+                        let mut add_defining_axiom = |name: Ident, lhs: &Expr, rhs: &Expr| {
+                            translate.push();
+                            translate.set_fuel_context(FuelContext::Head);
+                            let symbolic_lhs = translate.t_symbolic(lhs).into_dynamic(self);
+
+                            translate.set_fuel_context(FuelContext::Body);
+                            let symbolic_rhs = translate.t_symbolic(rhs).into_dynamic(self);
+
+                            axioms.push((
+                                name,
+                                translate.local_scope().forall(
+                                    &[&Pattern::new(self.ctx, &[&symbolic_lhs as &dyn Ast<'ctx>])],
+                                    &symbolic_lhs.smt_eq(&symbolic_rhs),
+                                ),
+                            ));
+                            translate.set_fuel_context(FuelContext::Call);
                             translate.pop();
+                        };
+
+                        if self.is_limited_function_decl(&func) {
+                            // fuel synonym axiom
+                            add_defining_axiom(func.name, &app, &app); // TODO: create a new name for the axiom
                         }
 
                         // create the axiom for the definition if there is a body
                         if let Some(body) = &*body {
-                            axioms.push((
-                                func.name, // TODO: create a new name for the axiom
-                                translate.t_bool(&builder.quant(
-                                    QuantOpKind::Forall,
-                                    func.inputs.node.iter().map(|param| param.name),
-                                    builder.binary(
-                                        BinOpKind::Eq,
-                                        Some(TyKind::Bool),
-                                        app,
-                                        body.clone(),
+                            // Defining axiom
+                            add_defining_axiom(func.name, &app, body); // TODO: create a new name for the axiom
+
+                            // Computing axiom
```

Comment:
Also here, capture in another method. And please add comments showing roughly what the generated axiom(s) look like.